core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -1599,4 +1696,31 @@ object KafkaZkClient {
       time, metricGroup, metricType)
     new KafkaZkClient(zooKeeperClient, isSecure, time)
   }
+
+
+  private def controllerZkVersionCheck(version: Int): Option[ZkVersionCheck] = {
+    if (version < 0)
+      None
+    else
+      Some(ZkVersionCheck(ControllerEpochZNode.path, version))
+  }
+
+  private def maybeThrowControllerMoveException(response: AsyncResponse): Unit = {
+    response.zkVersionCheckResult match {
+      case Some(zkVersionCheckResult) =>
+        val zkVersionCheck = zkVersionCheckResult.zkVersionCheck
+        if (zkVersionCheck.checkPath.equals(ControllerEpochZNode.path))
+          zkVersionCheckResult.opResult match {
+            case errorResult: ErrorResult =>
+              val errorCode = Code.get(errorResult.getErr)
+              if (errorCode == Code.BADVERSION)
+              // Throw ControllerMovedException when the zkVersionCheck is performed on the controller epoch znode and the check fails
+                throw new ControllerMovedException(s"Controller epoch zkVersion check fails. Expected zkVersion = ${zkVersionCheck.expectedZkVersion}")
+              else if (errorCode != Code.OK)
```

Comment:
Is it possible for error code to be `Code.OK` if `zkVersionCheckResult.opResult` is of type  `ErrorResult`?

Reply:
Yes. For example, if we wrap `check` + `create` in zookeeper `multi`, and `multi` fails  due to `create` fails, the result of `check` will be of type `ErrorResult` with `Code.OK` as error code.

Reply:
Cool. I see.